### ISSUE TITLE
Applying a function generic from a different context

### ISSUE THREAD

**Issue by ChuckJonas** [geoffreytools/free-types#2]
I ended up here from this [typescript issue.](https://github.com/microsoft/TypeScript/issues/1213) and think this library might solve the problem I'm running into.

I'm trying to create an library interface where a generic function can be passed in and the type basically resolved from a different context:

Hopefully this [Typescript-Playground](https://www.typescriptlang.org/play?jsx=0#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4EMxgA2qOAgnAL5wBmUEIcA5LRhgLQzoYDOjAUH06Y4AIQzVoGABK4AdgBNCGKAB4AShgCOAVx4wcAeW0wwxgHxwAvHAAUULQC44GnXoCUVi0ZPGA3AKEMOFIwYBUABTgMAA8YDAVuZwwAY2h5FW4YKGBZAHMcDVSodMzsvJwEe1dMpzkSOHtuSFluDFrZVAozMxwRKNj4+USxCXsZBSVVOpw67qtEPjg4ACNxSScRfyX8YCcERaW4AG0wXBgACzgcuABrDFQIajhwgF09g8PjkAwLiHkr2S3e6PZ4nM7nF5vWwfT6HKpOcJgi4vI7fX7yFGMKq6TKMF44GGwlZrexODQwbRQWRoTAqEQWOAAekZyTkNHs7ECiVWcG0rX+52UQRgEDg3HoGHOEAA7nAAET4IioOWg07I1E-KUYo5YrQ4mB4vCJC5BEaScaKZQA0wwAD8hMOHksFkRaohGvRmMazVahphFAOAYDAS4z3BiWs+yWjEZdGMGEY70+uR+ewaer0ewkECcpRyuQoOG9EBabTFWXzlC2lD4wfkKUIuHsNG0smSMGAJbgADdcIRgPIzkF1BnMmY7I4ktUYG4073+4O4vIyaP4BR-HxUi14DsnCEwuFwzhAiD5wOhxZIwdVqMy2fF9gDruFp85bGIPG5UnYSmYE4J5oOA3pIToWFGRLEreAB095Dv8TKMnAgRilK2iEP8PIILBS5ZhAOblmUBZUDYuCJMATzYRg8gAR40qkQ0rZuAIEHphSVK2MBzb0XUbgwX255LlB2YOgGnyiTWFBAA) isn't too hard to follow.

Sorry for such a direct question, please close/delete if this isn't appropriate.

**Comment by geoffreytools** [comment]
Hi,

No problem, I consider it part of the documentation.

People often mention higher kinded types when this use case comes up, but no implementation of HKT in TS can apply a generic function type with an argument dynamically, because the parameters in a generic function are not of the same kind as the parameters in a generic type alias or interface and TS doesn't provide any hook to access them at the type level.

However [type-lenses](https://github.com/geoffreytools/type-lenses) may help if you are ready to take a risk.

First, let me suggest something simpler that we can do at the value level, because it may be enough for your use case.

```typescript
declare function validate <Request>(req: Request): { validated: Request };

// we can actually do this at the value level
const validateFooString = validate<Paths['/route']['get']['request']>

const api: Api<Paths, typeof validateFooString> = {
  before: validate,
  api: {
    "/route": {
      get: (req, before) => {
        before.validated
         //        ^? (property) validated: { foo: string; }

        return (before as any).validated.foo
      }
    }
  }
}
```
[playground](https://www.typescriptlang.org/play?jsx=0#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4EMxgA2qOAgnAL5wBmUEIcA5LRhgLQzoYDOjAUH06Y4AIQzVoGABK4AdgBNCGKAB4AShgCOAVx4wcAeW0wwxgHxwAvHAAUULQC44GnXoCUVi0ZPGA3AKEMOFIwYBUABTgMAA8YDAVuZwwAY2h5FW4YKGBZAHMcDVSodMzsvJwEe1dMpzkSOHtuSFluDFrZVAozMxwRKNj4+USxCXsZBSVVOpw67qtEPjg4ACNxSScRfyX8YCcERaW4AG0wXBgACzgcuABrDFQIajhwgF09g8PjkAwLiHkr2S3e6PZ4nM7nF5vWwfT6HKpOcJgi4vI7fX7yFGMKq6TKMF44GGwlZrexODQwbRQWRoTAqERmQmHDyWCyI07I1E-c5-TGNZqtPFwGEUA4ikUBLjPcGJaz7JaMAD0dGMGEY70+uR+ewaWhxMD2EggTlKOVyFBwfIgLTacBNeUoW0ofHF8hShFw9ho2lkyRgwCtcAAbrhCMB5GcgupdXozHZHElqjA3Nrg6Hw3F5GTo5kHQJUi14KmwxGAGIQCAAZSypvmRfTGAi0qOiuVcTxzc1MHbWOzXZeDL4+ZzOycITC4WlOECILrpfLVbKuQssoOq1GNtncSwBxHC0+ACIlRAVfv1bDO0445ocGvJMyLHKicT1wA6TcYeSMw4KhVPuAAPQAfgEP97ApKlbFvT1cESOo3DfENiwzF9DUZEVPnQp0KCAA)

Now if we want to do this dynamically, an idea is to replace `unknown` in `ReturnType<B>` with `P[path][method]['request']`, which is not what we want semantically but yields the same result in this instance.

type-lenses require a path to replace some piece of type with another. We can write a utility type that looks for `unknown` in an arbitrarily nested object and generates the path to go there.

```typescript
// create a path leading to `unknown` in the object type
type CreatePath<T> = FilterPath<GetPaths<T>>

// get all possible paths and the value they lead to in the object type
type GetPaths<T> = T extends object
    ? { [K in keyof T]: [K, ...GetPaths<T[K]>] }[keyof T]
    : [T];

// exclude paths which don't lead to `unknown`
type FilterPath<T> = T extends [unknown, ...unknown[]]
  ? IsUnknown<Last<T>> extends true ? Init<T> : never
  : never;
```

With this path it is now possible to leverage `Replace`
```typescript
// in API
before: Replace<CreatePath<ReturnType<B>>, ReturnType<B>, P[path][method]['request']>
```

No change needs to be made to `api`
```typescript
const api: Api<Paths, typeof validate> = {
  before: validate,
  api: {
    "/route": {
      get: (req, before) => {
        before.validated
        //        ^? (property) validated: { foo: string; }

        return (before as any).validated.foo
      }
    }
  }
}
```

[playground](https://www.typescriptlang.org/play?jsx=0#code/JYWwDg9gTgLgBAbzgSQM4FUB2BrTEDumANHADICGqMJymw8AvnAGZQQhwDkrApjwLQwAnmB6pOAKFCRYiOACUeYADbkAxjzhNW7LsNH9lPTKjGSJ+zQCEezaDwAS5TABMjUADyKAjgFcx1HAA8r4wYKEAfHAAvHAAFFA83gBcCkn+VACUMVEhYaEA3BIWIpoAgmDAHgAKcDwAHjDGLqhpatAuHlRQwJgA5iSK7VCd3b0DiIl+AanOQiSJqJAmPLOYQgwRESRWdY3NrTZ2iU6u7h5zJHNbMYgScHAARrb2qVZFD+SVqQj3D3AAbTA5BgAAs4L04NgeEIIMw4NUALo-P7-QEgHhgiAuCGYKEwuEIoEg0GI5HxVFo-5TVLVYlgxEAjFYlyMzhTDIwTiIkiUqlPF6JVKKFTqHgeADCiRBPGqJK8mN8UEwABVSh4rFtBorlWrRBrtkTgQymZjQdi2RyAtyonAAPR2uAAdU05EScESorU4zgAANfDg8IRfXB8PRwXTjaTTSzLelrYjcWkYErVerNXz-tlolFIyTGczzayAeyxMtTNy4JSGH8azWSqI4FKeDK5WCPCqorEAGLAZRNKBt0EeADimKHqA7W2KljgY5gE6ntxVeyarlaEEeACseGoYJSAPxyAEAaST0Nh8JV5NPJAAdA-54uVafERFEwwARfCdfKakAb+DaaL2-Y8IO8qdsuq4HICAa4AQxBwA+d5wUGmAAmSfxHmgWDwYQHgUFQU5RA0a4tHAMBQP4cDYXQMBLqkmA8AAbmBfyMSxYFFEBCIkq0sS-A8nB2mwoQ8JwKJon0mI-B68ZUD8dgQKkYz9AwCxlhAKwqZRPoMB8WgSPWLi7qo7rMAGe7AFpcDMeQyjAC4MpwAq0xUBECRJMK8kwJksl2Q5TlNC43luYw3HtCY8BfMAqQVFUE4kJYhIBY5MpdncDzPMcqy2fZaVNEQfwxZJ-wAEQiRAYllaVUkyfEUwkNl9jZlEgn8gKOV3qlQU8C4mYPA6HVwAAegexTDYkKbKvEzXupQcBzJk3X5b1Lh3kpmY1mi22GQwQA)

Now the risk I mentioned is that `unknown` is a perfectly valid type which could appear multiple times in an object, and not necessarily in places you want to find/replace, so use this with caution!

**Comment by ChuckJonas** [comment]
First off, seriously thank you so much for taking the time to write out such a lengthly and complete response!  🙇 

> let me suggest something simpler that we can do at the value level, because it may be enough for your use case.

Unfortunately this isn't sufficient for my use case.  In my attempt to make the example code as simple as possible, I removed some critical context. The goal is to get the before response type to work across various routes, based on the request types of each route:

[Playground](https://www.typescriptlang.org/play?jsx=0#code/JYWwDg9gTgLgBAbzgFQJ5gKYBo4EMxgA2qOAgnAL5wBmUEIcA5LRhgLQzoYDOjAUH06Y4AIQzVoGABK4AdgBNCGKAB4AShgCOAVx4wcAeW0wwxgHxwAvHAAUULQC44GnXoCUVi0ZPGA3AKEMOFIwYBUABTgMAA8YDAVuZwwAY2h5FW4YKGBZAHMcDVSodMzsvJwEe1dMpzkSOHtuSFluDFrZVAozMxwRKNj4+USxCXsZBSVVOpw67qtEPjg4ACNxSScRfyX8YCcERaW4AG0wXBgACzgcuABrDFQIajhwgF09g8PjkAwLiHkr2S3e6PZ4nM7nF5vWwfT6HKpOcJgi4vI7fX7yFGMKq6TKMF44GGwlZrexODQwbRQWRoTAqERmQmHDyWCyI07I1E-c5-TGNZqtPFwGEUA4ikWCLjPcGJaz7JaMAD0dGMGAAjIx3p9cj89g0tDiYHsJBAnKUcrkKDg+RAWm04Ga8pQtpQsAdFcq4gAmDULLU6xB66qGxDLXBQU1Zc2WvVNG2tJzLCAQJRyJ2ivji+QpQhhoLUbSyZIwYA2uAAN1whGA8jOQXU+r0ZjsjiSQbcuorVZrcXkZIbmTTfFSLXgnertYAYkmAMqRx3WMfdjARaVHd0QFXqzHamB4tfYvR4hlDuPwHZOEJhcLSnCBEGLyczue5Cyyg6rUZ2h9xV3bUKaw4ACIlQ3OJVUAgC-WDZtNBwD9JGZCw5SJYlPwAOm-DB5EZQ4FQVFC4AAPQAfgEAj7ApKlbHg+w8ESOo3Awytxx7NDjUZEVPktGFgI9DBPQg5CoKcGC4JJDBEN9FCaIwJiu1rbCCLgPCCJInDlPw7huW0Qh-lWctmKXXsQzDCMygtdSKMpQEbBkui8A6RjMPkNikw44V0woIA)

It's going to take me a bit to fully grok this, but it definitely seems inline with what I'm looking for!

I do have one immediate followup question....

Why, if I replace the return type of validated with something that doesn't use the generic parameter `Response`, does the type `before` becomes the same as `request`?  

I assume it has something to do with the conditional types used here ` Replace<CreatePath<ReturnType<B>>, ReturnType<B>, P[path][method]['request']>` somehow looking for the type of `P[path][method]['request']` and falling back to `P[path][method]['request']` when it's not found?

[Playground](https://www.typescriptlang.org/play?jsx=0#code/JYWwDg9gTgLgBAbzgSQM4FUB2BrTEDumANHADICGqMJymw8AvnAGZQQhwDkrApjwLQwAnmB6pOAKFCRYiOACUeYADbkAxjzhNW7LsNH9lPTKjGSJ+zQCEezaDwAS5TABMjUADyKAjgFcx1HAA8r4wYKEAfHAAvHAAFFA83gBcCkn+VACUMVEhYaEA3BIWIpoAgmDAHgAKcDwAHjDGLqhpatAuHlRQwJgA5iSK7VCd3b0DiIl+AanOQiSJqJAmPLOYQgwRESRWdY3NrTZ2iU6u7h5zJHNbMYgScHAARrb2qVZFD+SVqQj3D3AAbTA5BgAAs4L04NgeEIIMw4NUALo-P7-QEgHhgiAuCGYKEwuEIoEg0GI5HxVFo-5TVLVYlgxEAjFYlyMzhTDIwTiIkiUqlPF6JVKKFTqHgeADCiRBPGqJK8mN8UEwABVSh4rFtBorlWrRBrtkTgQymZjQdi2RyAtyonAAPR2uAAdU05EScESorU4zgAANfDg8IRfXB8PRwXTjaTTSzLelrYjcWkYErVerNXz-tlolFIyTGczzayAeyxMtTNy4JSGH8azWSqI4FKeDK5WCPCqorEAGLAZRNKBt0EeADimKHqA7W2KljgY5gE6ntxVeyarlaEEeACseGoYJSAPxyAEAaST0Nh8JV5NPJAAdA-54uVafERFEwwARfCdfKakAb+DaaL2-Y8IO8qdsuq4HICAa4AQxBwA+d5wUGmAAmSfxHmgWDwYQHgUFQU5RA0a4tHAMBQP4cDYXQMBLqkmA8AAbmBfyMSxYFFEBCIkq0sS-A8nB2mwoQ8JwKJon0mI-B68ZUD8dgQKkYz9AwCxlhAKwqZRPoMB8WgSPWLi7qo7rMAGe7AFpcDMeQyjAC4MpwAq0xUBECRJMK8kwJksl2Q5TlNC4qSPBAEBGM4Wjce0JjwF8wCpBUVQTiQliEgFjkyl2dwPM8xyrLZ9lZU0RB-Alkn-AARCJEBiVVlVSTJ8RTCQ+X2NmUSCfyAoFXemVBTwLiZg8Do9XAAB6B7FONiQpsq8Tte6lBwHMmT9cVg0uHeSmZjWaL7YZDBAA)

**Comment by geoffreytools** [comment]
You're welcome.

The issue here is that `CreatePath<{ validated: boolean }>` returns `never`, because there is no path leading to `unknown`. The behaviour of `Replace` is unspecified when no path is provided.

I would compare `CreatePath<ReturnType<B>>` with `never` and branch over that

```typescript
before: CreatePath<ReturnType<B>> extends infer Path extends Query
  ? [Path] extends [never]
    ? ReturnType<B>
    : Replace<Path, ReturnType<B>, P[path][method]['request']>
  : never
```
Note that you need to import the constraint `Query` from type-lenses.

[playground](https://www.typescriptlang.org/play?jsx=0#code/JYWwDg9gTgLgBAbzgSQM4FUB2BrTEDumANHADICGqMJymw8AvnAGZQQhwDkrApjwLQwAnmB6pOAKFCRYiOACUeYADbkAxjxIBFAK48oQuE1bsuw0f2U9MqMZInmecAEI9m0HgAlymACZWoAB5FAEc9KhIAeR0YMBiAPjgAXjgACigeEIAuBUzwmABKZMTo2JiAbgkHEScAQTBgQIAFOB4ADxhrX1RctWhfQKooYEwAcxJFPqgBoZHxxAywsRgcnyESDNRIGx5VzCEGePiSZ1aOrp7XdwzvPwDAtZI1o+TECTg4ACM3DxznSo+5AaOQQ7w+cAA2mByDAABZwEZwbA8IQQZhwJoAXRBYPBkJAPDhEF8CMwSJRaIxUJhsMx2LSuLx4MWOSa1LhmIhBKJvk5nEW+U4mJIjKZXx+GRyAGEMjCeE0acFCTooJgACo1QLOF7tTp+HojZj6DE0s567pwXT6ISisVwAD8kIVHLNF0hmB4ADd9JjbXaHbkYCr1ZrtX67TlFCp1DxmjSJsrVRrRFrjlToRyuYTYcS+QLlkL4uG8TkPd6oOGiklEmyM7Sszy82JtrYhXBGQwwZ3O9VRHAZTw5c7YYE1YkUgAxYDKTpQYeBADihOHqFHRyqjjgS5gK7XrzVrv1cAgnwAVjw1DBGY6kBCANKk8mo9Fq+n3kgAOi-293avvmPiTEjAhZFnzgV9GRyCEIN7JwpxnfR5zHfdDwtCEdBwPBCE-L8MNwAhMAhOkwUdNAsHwwhAgoKg10SXU3RgKA9ADWh6D3UsvX0MEOPLSpYJNOEehSUEPk4AB6NgYh4TgcTxUZCRBOB8yoEF3AgHJZjGBgNmbCAdg0xi5iMAEjAkHtfAvVQMhYDDL2APS4E9chlGAXw5TgJUlioeJ0kySM8mWApFKcly3M6Xwck+CAICsHxjKqPobHgIFgByepGhXEhHEpELXLlcc3g+b5rl2RznLyzoiDBFLZPBAAiCSICkurarkhS0kWEhio8KtEhEu1uoyD9crCnhfHDMSxP9AA9e0qn9DIg1VNJBqcSg4DWAphvK0bfA-NTbU7PEjtMhggA)

This reminds me that when using `ReturnType` on a generic function, if the generic has a type constraint, it's going to be substituted with the type constraint instead of `unknown`, and as a result `CreatePath` won't detect any path either. One way to deal with this could be to look for supertypes of `P[path][method]['request']>` instead of `unknown` when generating the path, but as you can imagine the risk of collision with legitimate types becomes very high.

**Comment by ChuckJonas** [comment]
Awesome thanks! 

I've been messing around with `type-lenses` for the last hour and starting to get it.  Still TBD if I'll have the skills to incorporate this back into the full vision, but such a cool little utility library!  

I think I enjoy writing types more than the code itself 😅

**Comment by geoffreytools** [comment]
Yeah that's something to be wary of actually ;)

By the way, this issue made me consider including path finding to type-lenses, and changing the behaviour of `Replace` when the path is `never` so that users don't need to branch.

Don't hesitate to post issues if the documentation wasn't very clear. I don't get a lot of feedback.

**Comment by geoffreytools** [comment]
type-lenses now implements `FindReplace` for this kind of use case, so this:

```typescript
before: CreatePath<ReturnType<B>> extends infer Path extends Query
  ? [Path] extends [never]
    ? ReturnType<B>
    : Replace<Path, ReturnType<B>, P[path][method]['request']>
  : never
```
can be re-written as:
```typescript
before: FindReplace<ReturnType<B>, unknown, [P[path][method]['request']]>
```
> Notice the brackets around `P[path][method]['request']`. `FindReplace` expects a tuple of replace values (or a replace callback)


`Replace` now fails gracefully when the query is `never`, and a general `FindPath` utility type was added to the library, so the following would also work:
```typescript
before: Replace<FindPath<ReturnType<B>, unknown>, ReturnType<B>, P[path][method]['request'], any>
```
> Notice the `any` at the end. `Replace` now type checks the replace value against the query, which is problematic when it requires a generic to be resolved (here `B`). `any` lets us disable the check.
